### DISCUSSION TITLE
Displaying Real-Time Data [Newbie]

### DISCUSSION THREAD

**Discussion by natabma** [patrykandpatrick/vico#602] [Questions] [answered]
I want to display real-time Data with a Line Chart. More specifically, i have a sensor that collects EKG Data and i want to create an EGK Diagram, like so: 
![image](https://github.com/patrykandpatrick/vico/assets/77214481/ab4a5f59-a4b3-46a3-af34-5b93dc4609c6)

I tried to implement this beautiful library like so: The EKG data is collected in an 1-D Array that is kept at a length of 200 and is continously updated. (This works, i checked the logs.) The Problem is, that the chart is not updated while the data is collected. **It only updates, when i stop the capture.** I tried my best, but to no avail. I found a function called `registerForUpdates`, which sounded right to me - but I struggle to implement it. Could someone give me some tips?

 My implementation looks like this:

![image](https://github.com/patrykandpatrick/vico/assets/77214481/4922d4a1-0207-4677-a56a-36ca4bd4cab0)

**Comment by patrickmichalik** [comment, marked as answer]
Hello! Does the issue persist when you set `diffAnimationSpec` (which is a parameter of `CartesianChartHost`) to `null`? Also, given the high transaction frequency, I’d suggest using `runTransaction` instead of `tryRunTransaction`.

**Reply by natabma** [reply]
Dear Patrick,
thank you so much for your quick reply. Indeed, setting `diffAnimationSpec` to `null` was the solution. I also switched from `tryRunTransaction` to `runTransaction`.
Props btw for this Library and especially how well everything is documented. It is a joy to use it.

**Reply by patrickmichalik** [reply]
I’m glad to hear that this solved the issue, @natabma. Thanks for the kind words!

I wanted to share one more observation. While reading the content of the array and running a `Transaction` every 100 ms will keep the chart up to date, it can also lead to unnecessary operations being performed. Every 100 ms, regardless of whether the array’s content has changed, a list is created and a `Transaction` is run. As I understand, the data may stop being updated at some point. From then on, the same data will be unnecessarily transformed ten times per second, and the chart will be needlessly redrawn.

It’s preferable to run `Transaction`s only when needed. The optimal approach depends on the specifics of the use case. Usually, you can move the `CartesianChartModelProducer` to the screen’s `ViewModel` and trigger `Transaction`s from a callback linked to the data source. Using `Flow`s can be helpful.

In [Vico 2.0.0 Alpha 11](https://github.com/patrykandpatrick/vico/releases/tag/v2.0.0-alpha.11), we’ve optimized `CartesianChartModelProducer` to skip `Transaction`s that don’t change anything. It’s still preferable to avoid requesting unnecessary `Transaction`s, but such `Transaction`s are no longer that much of a concern. I’d recommend updating.

**Reply by natabma** [reply]
I just wanted to get back to this.
You really hepled me a lot with your suggestions. It runs quite smooth now. 11/10 Support

**Reply by patrickmichalik** [reply]
I’m glad to hear that, @natabma. Thanks for the kind words!
